feat(VForm): add form level disabled prop

Logikgate opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://vuetifyjs.com/getting-started/contributing

Testing and markup sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

<!-- We use conventional-changelog-angular for all commit structures -->
<!-- https://vuetifyjs.com/getting-started/contributing#commit-guidelines-w-commitizen -->


## Description
A disabled VForm now disables all child inputs. This fixes #1707.
This change was implemented so that the child inputs could still have their own disabled props but the top level VForm disabled prop takes precedence over the child input disabled state. 

<!-- Describe your changes in detail -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #4213 or fixes #2312 -->

## Motivation and Context
It is common to want all of the inputs on the form to be disabled all at once.
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->


## How Has This Been Tested?
Visually tested with all components that inherit from VInput
Added unit test to check VTextfield
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | e2e | none -->

## Markup:
<!-- Information on how to setup your local development environment can be found here: -->
<!-- https://vuetifyjs.com/getting-started/contributing#setup-dev-environment -->

<!-- Paste markup for testing your change --->
```
<v-form disabled>
      <v-text-field></v-text-field>
</v-form>
```

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)
###

KaelWD commented:
Related: #2033

As discussed there I'd prefer this done with provide/inject rather than updating data imperatively.

Logikgate commented:
@KaelWD Ah yes I missed that conversation in that PR. I'm willing to refactor this if you provide me a little bit of guidance. 

The way this is working right now is it is piggybacking on the provide/inject of the validatable mixin.  I can make make a separate mixin to do the same thing but I'm unsure how that will change the implementation of form level disabled prop being able to override the input's disabled prop.  Or are you thinking that it would work in the opposite direction? i.e. the VForm would inject into the VInputs

KaelWD commented:
See VExpansionPanel for example. VForm provides itself and the controls have a `isDisabled` computed property that checks `this.form.disabled`.

KaelWD commented on packages/vuetify/src/mixins/registrable/index.ts:
```diff
@@ -15,10 +16,10 @@ export type Registrable<T extends string, C extends VueConstructor | null = null
 export function inject<
   T extends string, C extends VueConstructor | null = null
-> (namespace: T, child?: string, parent?: string): Registrable<T, C> {
+> (namespace: T, child?: string, parent?: string, nestingOptional?: boolean): Registrable<T, C> {
```
Child/parent are only used for the warning, if it's optional you just don't use those arguments.

Logikgate replied:
Got it. I did this because inputs are often used without a v-form so most users will see a warning that states a component "must" be used inside a v-form so even though it's a warning it reads more like an error which is misleading. Also without this 322 tests fail.

I'm good with removing this and updating all of the tests to mock the provide but I just want to make sure showing this warning to everyone will be desired behavior. Maybe the warning could be reworded to read less like an error?

Logikgate replied:
Thanks I missed that. Fixed now 👍

Logikgate commented:
@KaelWD I am updating the test to cover all of the components that should be set to disabled if the form is disabled.  There are two inputs that do not inherit from v-input and therefore do not get disabled by the form level input:

VRadio: This only matters if it was used without a group but using it without a group doesn't sense so maybe it doesn't matter. It looks like if you use a VRadio without a VRadioGroup no warning is shown though, should I adjust this so the warning shows?

https://github.com/vuetifyjs/vuetify/blob/d8b3adc45d6146e7351b3607612d9b889610e309/packages/vuetify/src/components/VRadioGroup/VRadio.ts#L29

VSimpleCheckbox: Should this inherit from v-input? Or should I just do the provide/inject to make the form level disable work?

https://github.com/vuetifyjs/vuetify/blob/d8b3adc45d6146e7351b3607612d9b889610e309/packages/vuetify/src/components/VCheckbox/VSimpleCheckbox.ts#L12

KaelWD commented:
> should I adjust this so the warning shows?

I think that was intentional, but yeah it probably doesn't make sense to use it in a form without radio-group. 

> VSimpleCheckbox

Not supposed to be a full replacement for VCheckbox, it's for performance critical stuff like data table selection. Oh also it's functional so that won't work anyway, just ignore it.

johnleider commented on packages/vuetify/src/components/VForm/VForm.ts:
```diff
@@ -25,11 +25,18 @@ export default mixins(
+
   inheritAttrs: false,
 
   props: {
```
sort children alphabetically

johnleider commented on packages/vuetify/src/components/VForm/VForm.ts:
```diff
@@ -25,11 +25,18 @@ export default mixins(
 
+  provide (): object {
+    return {
+      form: this,
```
make this one line with the return

johnleider commented on packages/vuetify/src/components/VForm/__tests__/VForm.spec.ts:
```diff
@@ -1,27 +1,58 @@
+import Vuetify from '../../../framework'
+
+// Components
+import VForm from '../VForm'
```
Is there anything of value added by testing all of the input components? All validation should be handled in the validatable mixin.

Logikgate commented:
@johnleider thanks for the review! I incorporated your comments; moved the isDisabled computed prop to the Registratable mixin, VForm test now only tests one component and fixed the couple of formatting issues

johnleider commented:
Please resolve conflicts.

johnleider commented on packages/vuetify/src/components/VForm/__tests__/VForm.spec.ts:
```diff
@@ -1,28 +1,49 @@
-    })
-  },
-})
+Vue.use(Vuetify)
```
What is the purpose of this?

Logikgate replied:
It came in via the dev branch merge. The wait function is used on line 178.

johnleider commented on packages/vuetify/src/components/VForm/__tests__/VForm.spec.ts:
```diff
@@ -245,4 +266,28 @@ describe('VForm.ts', () => {
   })
+
+  it('should disable all inputs', async () => {
+    const inputs = [
```
move to one line

johnleider commented on packages/vuetify/src/components/VForm/__tests__/VForm.spec.ts:
```diff
@@ -245,4 +266,28 @@ describe('VForm.ts', () => {
+    ]
+
+    const wrapper = mountFunction({
+      propsData: {
```
move to one line

johnleider commented on packages/vuetify/src/components/VForm/__tests__/VForm.spec.ts:
```diff
@@ -245,4 +266,28 @@ describe('VForm.ts', () => {
+      propsData: {
+        disabled: true,
+      },
+      slots: {
```
move to one line

jacekkarczmarczyk commented on packages/docs/src/lang/en/components/Forms.json:
```diff
@@ -23,6 +23,7 @@
     }
   },
   "props": {
+    "disabled": "If enabled this will set all child inputs to disabled",
```
Current message ("if enabled ... to disabled") sounds weird to me, maybe "If set to `true`" or just "Disables all child inputs"
